src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,238 @@
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(T value, Node<T> next, Node<T> prev) {
+            this.item = value;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void linkBefore(T value, Node<T> next) {
+        Node<T> previousNode = next.prev;
+        Node<T> newNode = new Node<>(value, next, previousNode);
+        next.prev = newNode;
+        if (previousNode == null) {
+            head = newNode;
+        } else {
+            previousNode.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkNewHead(T value) {
+        Node<T> first = head;
+        Node<T> newNode = new Node<>(value, head, null);
+        head = newNode;
+        if (first == null) {
+            tail = newNode;
+        } else {
+            first.prev = newNode;
+        }
+        size++;
+    }
+
+    private void linkNewTail(T value) {
+        Node<T> newNode = new Node<>(value, null, tail);
+        tail.next = newNode;
+        tail = newNode;
+        size++;
+    }
+
+    private T unlink(Node<T> nodeToUnlink) {
+        final T removeValue = nodeToUnlink.item;
+        Node<T> nextNode = nodeToUnlink.next;
+        Node<T> previous = nodeToUnlink.prev;
+        if (previous == null) {
+            head = nextNode;
+        } else {
+            previous.next = nextNode;
+            nodeToUnlink.prev = null;
+        }
+
+        if (nextNode == null) {
+            tail = previous;
+        } else {
+            nextNode.prev = previous;
+            nodeToUnlink.next = null;
+        }
+        nodeToUnlink.item = null;
+        size--;
+        return removeValue;
+    }
+
+    private T unlinkHead(Node<T> nodeToUnlink) {
+        final T removeItem = nodeToUnlink.item;
+        Node<T> nextNode = nodeToUnlink.next;
+        nodeToUnlink.item = null;
+        nodeToUnlink.next = null;
+        head = nextNode;
+        if (nextNode == null) {
+            tail = null;
+        } else {
+            nextNode.prev = null;
+        }
+        size--;
+        return removeItem;
+    }
+
+    private T unlinkTail(Node<T> nodeToUnlink) {
+        final T removeItem = nodeToUnlink.item;
+        Node<T> prev = nodeToUnlink.prev;
+        nodeToUnlink.item = null;
+        nodeToUnlink.prev = null;
+        tail = prev;
+        if (prev == null) {
+            head = null;
+        } else {
+            prev.next = null;
+        }
+        size--;
+        return removeItem;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE + index);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(INDEX_OUT_OF_BOUNDS_MESSAGE + index);
+        }
+    }
```

Comment:
You use these methods only once and they are short. Not sure you need them

Reply:
Disagree, I guess in this way I make code more readable.
Also, I had such kind of implementation for ArrayList.
Left for mentor decision mb.